src/renderer/api/kube-watch-api.ts

```diff
@@ -1,201 +1,326 @@
+  protected processBuffer(events: string[]): string {
+    for (const json of events) {
+      try {
+        const kubeEvent: IKubeWatchEvent = JSON.parse(json);
+        const message = this.getMessage(kubeEvent);
+
+        this.onMessage.emit(message);
+      } catch (error) {
+        return json;
       }
     }
-  }
 
-  protected writeLog(...data: any[]) {
-    if (isDevelopment) {
-      console.log("%cKUBE-WATCH-API:", `font-weight: bold`, ...data);
-    }
+    return "";
   }
 
-  addListener(store: KubeObjectStore, callback: (evt: IKubeWatchEvent) => void) {
-    const listener = (evt: IKubeWatchEvent<KubeJsonApiData>) => {
-      if (evt.type === "ERROR") {
-        return; // e.g. evt.object.message == "too old resource version"
-      }
+  protected getMessage(event: IKubeWatchEvent): IKubeWatchMessage {
+    const message: IKubeWatchMessage = {};
+
+    switch (event.type) {
+      case "ADDED":
+      case "DELETED":
 
-      const { namespace, resourceVersion } = evt.object.metadata;
-      const api = apiManager.getApiByKind(evt.object.kind, evt.object.apiVersion);
+      case "MODIFIED": {
+        const data = event as IKubeWatchEvent<KubeJsonApiData>;
+        const api = apiManager.getApiByKind(data.object.kind, data.object.apiVersion);
 
-      api.setResourceVersion(namespace, resourceVersion);
-      api.setResourceVersion("", resourceVersion);
+        message.data = data;
 
-      ensureObjectSelfLink(api, evt.object);
+        if (api) {
+          ensureObjectSelfLink(api, data.object);
 
-      if (store == apiManager.getStore(api)) {
-        callback(evt);
+          const { namespace, resourceVersion } = data.object.metadata;
+
+          api.setResourceVersion(namespace, resourceVersion);
+          api.setResourceVersion("", resourceVersion);
+
+          message.api = api;
+          message.store = apiManager.getStore(api);
+        }
+        break;
       }
-    };
 
-    this.onData.addListener(listener);
+      case "ERROR":
+        message.error = event as IKubeWatchEvent<KubeJsonApiError>;
+        break;
 
-    return () => this.onData.removeListener(listener);
+      case "STREAM_END": {
+        this.onServerStreamEnd(event as IKubeWatchEventStreamEnd);
+        break;
+      }
+    }
+
+    return message;
   }
 
-  reset() {
-    this.subscribers.clear();
+  protected async onServerStreamEnd(event: IKubeWatchEventStreamEnd) {
+    const { apiBase, namespace } = parseKubeApi(event.url);
+    const api = apiManager.getApi(apiBase);
+
+    if (api) {
+      try {
+        await api.refreshResourceVersion({ namespace });
+        this.connect();
+      } catch (error) {
+        this.log({
+          message: new Error(`Failed to connect on single stream end: ${error}`),
+          meta: { event, error },
+        });
+
+        if (this.isActive) {
+          setTimeout(() => this.onServerStreamEnd(event), 1000);
+        }
+      }
+    }
```

Comment:
IMO less indentation is generally easier to read.

```suggestion
    if (!api) {
      return;
    }
    
    try {
      await api.refreshResourceVersion({ namespace });
      this.connect();
    } catch (error) {
      this.log({
        message: new Error(`Failed to connect on single stream end: ${error}`),
        meta: { event, error },
      });

      if (this.isActive) {
        setTimeout(() => this.onServerStreamEnd(event), 1000);
      }
    }
```